api/coda.py

```diff
@@ -142,76 +127,291 @@ def update_user_stamps(self, user: DiscordUser, stamp_count: float) -> None:
     #   Questions   #
     #################
 
-    def get_question_row(self, question_id: str) -> QuestionRow:
-        """Get QuestionRow by its ID"""
+    def update_questions_cache(self) -> None:
+        """Update the cache of the
+        [All answers coda table](https://coda.io/d/AI-Safety-Info_dfau7sl2hmG/All-Answers_sudPS#_lul8a).
+
+        Gets called during initialization and every ~10 minutes by Questions module.
+        """
+        # get coda table
+        questions = self.doc.get_table(self.ALL_ANSWERS_TABLE_ID)
+        # parse its rows
+        question_rows = [parse_question_row(row) for row in questions.rows()]
+        # convert into dataframe
+        self.questions_df = pd.DataFrame(question_rows).set_index("id", drop=False)
```

Comment:
Thanks. I replaced updating every 10 minutes and instead wrote a method which doesn't overwrite the current dataframe but (1) updates its rows, so that they are up-to-date with coda (2) deletes rows which are not in coda (3) adds new rows that are in coda but are missing from the DF.

"Dropping" the table from test coda, updating Stampy's cache (which resulted in emptying the DF), bringing the table on coda back, and updating cache again (repeat a few times) didn't impact memory used by the process at all, whereas the old one did, so I think this one is safer. I'm still keeping the old one (can be called by `s, hardreload questions` whereas the new one can be called by `s, refresh questions` and several others) in case the new method doesn't work for some reason and people will need the cache to be up-to-date with coda.

The new method also gets called when Stampy tries to access a question by GDoc link which is missing from his current cache.